src/pubsub.c

```diff
@@ -271,14 +271,16 @@ int pubsubSubscribeChannel(client *c, robj *channel, pubsubtype type) {
             slot = getKeySlot(channel->ptr);
         }
 
-        de = kvstoreDictAddRaw(*type.serverPubSubChannels, slot, channel, &existing);
+        robj *channel_copy = dupStringObject(channel); /* Copy the channel name to avoid references
+                                                        * that could prevent memory defragmentation. */
```

Comment:
> but maybe it's easy to change it so that they all share the same robj (i.e. take the one from the kvstore and re-use it when adding a client). 

good idea, tested with the above python script, memory usage drop from 2.8G to 1G.

Reply:
> then when defragging the kvstore, we should update all the clients.

if so, we should iterator over all clients to update channel after defragging the channel in the kvstore.
that is, when there are 10,000 clients, updating a channel requires traversing these 10,000 clients.

Reply:
great. let's be sure to mention this memory optimization in the top comment (then mark this one as resolved)

Reply:
Updated